viam-cartographer/src/carto_facade/carto_facade.cc

```diff
@@ -627,8 +627,46 @@ int CartoFacade::Stop() {
     return VIAM_CARTO_SUCCESS;
 };
 
-int CartoFacade::AddSensorReading(viam_carto_sensor_reading *sr) {
-    return VIAM_CARTO_SUCCESS;
+void CartoFacade::AddSensorReading(const viam_carto_sensor_reading *sr) {
+    std::string sensor = to_std_string(sr->sensor);
+    if (config.sensors[0] != sensor) {
+        VLOG(1) << "expected sensor: " << sensor << " to be "
+                << config.sensors[0];
+        throw VIAM_CARTO_SENSOR_NOT_IN_SENSOR_LIST;
+    }
+    std::string sensor_reading = to_std_string(sr->sensor_reading);
+    if (sensor_reading.length() == 0) {
+        throw VIAM_CARTO_SENSOR_READING_EMPTY;
+    }
+
+    int64_t sensor_reading_time_unix_micro = sr->sensor_reading_time_unix_micro;
+    auto [success, measurement] =
+        viam::carto_facade::util::carto_sensor_reading(
+            sensor_reading, sensor_reading_time_unix_micro);
+    if (!success) {
+        throw VIAM_CARTO_SENSOR_READING_INVALID;
+    }
+
+    cartographer::transform::Rigid3d tmp_global_pose;
+    bool update_latest_global_pose = false;
+
+    if (map_builder_mutex.try_lock()) {
+        map_builder.AddSensorData(measurement);
+        auto local_poses = map_builder.GetLocalSlamResultPoses();
+        if (local_poses.size() > 0) {
+            update_latest_global_pose = true;
+            VLOG(1) << "local_poses.size(): " << local_poses.size();
+            tmp_global_pose = map_builder.GetGlobalPose(local_poses.back());
+        }
+        map_builder_mutex.unlock();
+        if (update_latest_global_pose) {
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_global_pose = tmp_global_pose;
+        }
+        return;
+    } else {
+        throw VIAM_CARTO_UNABLE_TO_ACQUIRE_LOCK;
+    }
```

Comment:
Will this function be called during offline mode?

Reply:
yes